doc/man/gnunet-config.1
.\" This file is part of GNUnet.
.\" Copyright (C) 2001-2019 GNUnet e.V.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the file
.\" FDL-1.3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/fdl.html}.
.\"
.\" Alternately, this document is also available under the General
.\" Public License, version 3 or later, as published by the Free Software
.\" Foundation.  A copy of the license is included in the file
.\" GPL3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/gpl.html
.\"
.\" SPDX-License-Identifier: GPL3.0-or-later OR FDL1.3-or-later
.\"
.Dd July 15, 2012
.Dt GNUNET-CONFIG 1
.Os
.Sh NAME
.Nm gnunet-config
.Nd manipulate GNUnet configuration files
.Sh SYNOPSIS
.Nm
.Op Fl b Ar BACKEND | Fl -supported-backend= Ns Ar BACKEND
.Op Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
.Op Fl f | -filename
.Op Fl h | -help
.Op Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
.Op Fl o Ar OPTION | Fl -option= Ns Ar OPTION
.Op Fl S |-list-sections
.Op Fl s Ar SECTION | Fl -section= Ns Ar SECTION
.Op Fl V Ar VALUE | Fl -value Ar VALUE
.Op Fl v | -version
.Op Fl W | -rewrite
.Sh DESCRIPTION
.Nm
can be used to read or modify GNUnet configuration files.
.Bl -tag -width indent
.It Fl b Ar BACKEND | Fl -supported-backend= Ns Ar BACKEND
Tests whether the specified BACKEND is supported by the current installation.  The backend must match the name of a plugin, i.e. "namestore_postgres" for the Postgres database backend of the "NAMESTORE" service. If the BACKEND is supported, gnunet-config will return a status code of 0 (success), otherwise 77 (unsupported).  When this option is specified, no other options may be specified. Specifying this option together with other options will cause gnunet-config to return a status code of 1 (error).
.It Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
Use the configuration file FILENAME.
.It Fl f | -filename
Try to perform expansions as if the option values represent filenames (will also be applied even if the option is not really a filename).
.It Fl h | -help
Print short help on options.
.It Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
Use LOGLEVEL for logging.
Valid values are DEBUG, INFO, WARNING and ERROR.
.It Fl o Ar OPTION | Fl -option= Ns Ar OPTION
Which configuration option should be accessed or edited.
Required to set a value.
If not given, all values of a given section will be printed in the format "OPTION = VALUE".
.It Fl S |-list-sections
List available configuration sections for use with --section.
.It Fl s Ar SECTION | Fl -section= Ns Ar SECTION
Which configuration section should be accessed or edited.
Required option.
.It Fl V Ar VALUE | Fl -value Ar VALUE
Configuration value to store in the given section under the given option.
Must only be given together with -s and -o options.
.It Fl v | -version
Print GNUnet version number.
.It Fl W | -rewrite
Consider differences to defaults only.
.El
.Sh SEE ALSO
The full documentation for gnunet is maintained as a Texinfo manual.
If the
.Xr info 1
and gnunet programs are properly installed at your site, the command
.Pp
.Dl info gnunet
.Pp
should give you access to the complete handbook,
.Pp
.Dl info gnunet-c-tutorial
.Pp
will give you access to a tutorial for developers.
.sp
Depending on your installation, this information is also available in
.Xr gnunet 7 and
.Xr gnunet-c-tutorial 7 .
.\".Sh HISTORY
.\".Sh AUTHORS
.Sh BUGS
Report bugs by using
.Lk https://bugs.gnunet.org
or by sending electronic mail to
.Aq Mt gnunet-developers@gnu.org .


ChangeLog
Thu Oct 17 00:00:00 UTC 2019
  Added support for doas, use it in some places conditionally
  if sudo is not present. Fixed sudo detection. -ng0

Mon Oct 14 00:00:00 UTC 2019
  Added 'pretty' make rule to run uncrustify over the source tree.
  -ng0

Mon Sep 16 00:00:00 UTC 2019
  Remove checks for CXX and CPP. Scans found no
  (remaining) cxx code. ccache with clang on
  NetBSD 9 hit no rebuilds. -ng0

Mon Sep 10 00:00:00 UTC 2019
  Drop win32 and mingw support after discussion with
  Christian. It has been unmaintained in our code base
  for years and there are no known users.
  Future implementors for Windows 10 support can
  refer to this commit as reference. -ng0

Mon Sep 9 00:00:00 UTC 2019
  Remove plibc, closes #5877 as discussed in
  https://lists.gnu.org/archive/html/gnunet-developers/2019-09/msg00002.html
  -ng0

Sun Sep 1 13:25:07 2019 +0200
  Improved support for external applications/services wrt
  localization/gettext.
  https://lists.gnu.org/archive/html/gnunet-developers/2019-09/msg00000.html

Tue 20 Aug 2019 04:32:09 PM CEST
  Fixed fatal bug in gnunet-bcd that prevented the main page
  from showing up. -CG

Mon Aug 19 17:36:48 2019 +0200
  REST: Added API plugin to read/write GNUnet configuration options -alexia

Sun Jul 28 00:00:00 2019 +0000
  Adding --enable-texinfo4 configure switch, defaulting to off.
  This allows distributors to override the texinfo 5.x and later
  commandline options we use, and allows us on the serverside to
  build html with newer output options. -ng0

Wed 24 Jul 2019 05:43:43 PM CEST
  Releasing GNUnet 0.11.6. -CG

Mon Jul 15 09:57:20 2019 +0200
  The REST server now has a configurable CORS Origin setting -schanzen

Fri Jul 5 22:11:04 2019 +0200
  Added detection of SUID binaries
  through GNUNET_OS_get_suid_binary_path -lurchi

Sat 29 Jun 2019 10:18:55 AM CEST
  Renamed "heap" namestore backend to "flat" as it does
  persist data in a file. Made it work with databases larger
  than 16 MB. -CG

Fri Jun 28 14:10:53 2019 +0200
  RECLAIM: Reissue of tickets (same attributes, same identity)
  will no longer create a new ticket, but reuse the old. -alexia

Thu Jun 27 10:34:46 2019 +0200
  The REST server now returns the Origin header value in the
  CORS header reponse. -schanzen

Thu Jun 27 09:08:06 2019 +0200
  GNS NSS plugin will not timeout and not hang forever and block
  gethostbyname. Final resolution pending #5782 -schanzen

Sat Jun 22 10:38:43 2019 +0200
  Add gnutls support for GNS CA certificate generation -schanzen

Sat 15 Jun 2019 11:45:35 PM CEST
  Improved gnunet-gns performance for many ego scenario. -CG

Fri 14 Jun 2019 07:17:40 PM CEST
  Add option to gnunet-identity to display private keys. -CG

Wed Jun 5 21:19:46 2019 +0200
 Releasing GNUnet 0.11.5. -schanzen

Wed May 29 11:17:49 2019 +0200
  Added CAA record type (RFC 6844). -schanzen

Sat 25 May 2019 08:27:39 AM CEST
  Significantly speed-up ego creation and deletion via
  gnunet-identity command when there are many (hundreds)
  of egos at the same peer. -CG

Sun May 12 12:00:00 2019 +0200
  Releasing GNUnet 0.11.4. -MS

Thu May 9 23:35:24 2019 +0200
  gnunet-identity can now output the public key on creation of new
  egos, and -e can act as a filter when listing all egos. -CG

Thu May 2 08:44:37 2019 +0200
  gnunet-arm now logs to file by default -CG

Sat Apr 27 21:50:57 2019 +0200
  RECLAIM: Added ECDH encryption for authorization code payload -schanzen

Sat 27 Apr 2019 06:36:31 AM CEST
  Bump libmicrohttpd version requirement as some intermediate
  versions have relevant bugs that cause problems. -CG

Sat Apr 27 01:50:43 2019 +0200
  RECLAIM: No longer experimental -schanzen

Fri Apr 26 11:08:21 2019 +0200
  REST: Configuration plugin added -schanzen

Tue Apr 23 23:00:00 UTC 2019
  All manpages are properly licensed. -ng0

Sun Apr 21 22:22:22 UTC 2019
  All manpages are now in mdoc format. -ng0

Sat Apr 20 18:38:43 UTC 2019
  Remove optional gnunet-download-manager.scm and
  with it the optional dependency on a no longer
  maintained version of GNU Guile (#5682). -ng0

Sun Apr 14 09:54:20 2019 +0200
  RECLAIM: Moved away from ABE to GNS encryption. Minor API changed and
  renaming of records. Deprecated SQLite plugin in favor of namestore.
  -schanzen

Sun Apr 7 14:40:59 2019 +0200
  Releasing GNUnet 0.11.3. -CG

Sat 06 Apr 2019 02:36:17 PM CEST
  Introducing GNUNET_memcmp() and GNUNET_is_zero() for better typing
  when comparing memory areas. -CG

Fri 05 Apr 2019 11:38:53 PM CEST
  Use paths of sysctl, ip, ifconfig and ip(6)tables from configure
  if they work. (#5352). -CG

Thu 04 Apr 2019 12:39:44 PM CEST
  Hotfix for build issue. Releasing GNUnet 0.11.2. -CG/MS

Wed 03 Apr 2019 03:18:43 PM CEST
  Releasing GNUnet 0.11.1. -CG

Wed 03 Apr 2019 01:54:44 PM CEST
  Merged gnunet-qr-c-2 branch, replacing Python-based implementation
  of gnunet-qr with C-based implementation that also integrates
  gnunet-uri. -HG/MS/CG

Wed 03 Apr 2019 01:15:57 PM CEST
  Fixed a few test case regressions (GNS, FS). -CG

Mon Apr 1 22:00:00 UTC 2019
  documentation/buildsystem: manpages and other documentation parts
  can now be build more controlled again, this fixes the regression
  in 0.11.0 where documentation was an on/off switch.
  New introduced: --enable-include-manpages.
  Fixed: texi2mdoc integration (--enable-texi2mdoc-generation).
  buildsystem: decrease python dependency to 3.4 from
  previously 3.7.
  -ng0

Tue Mar 12 10:06:17 CET 2019
  REST: Move back from rest-plugin into respective subsystem directories.

Mon Mar 11 02:47:55 CET 2019
  Make configure complain if no working libidn is detected. -CG

Sat Mar 9 15:58:45 2019 +0100
  REST: Config option for address bind. Defaults to localhost.
  GNS: dns2gns/gns-proxy config option for address bind. Defaults to localhost.

Sat Mar 9 01:58:22 CET 2019
  gnunet-publish now by default does not expose the creation time,
  option -d no longer exists, to enable the old behavior use -E. -CG

Sat Mar 2 14:23:10 2019 +0100

src/util/gnunet-config.c
/*
     This file is part of GNUnet.
     Copyright (C) 2012 GNUnet e.V.

     GNUnet is free software: you can redistribute it and/or modify it
     under the terms of the GNU Affero General Public License as published
     by the Free Software Foundation, either version 3 of the License,
     or (at your option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     Affero General Public License for more details.

     You should have received a copy of the GNU Affero General Public License
     along with this program.  If not, see <http://www.gnu.org/licenses/>.

     SPDX-License-Identifier: AGPL3.0-or-later
 */

/**
 * @file util/gnunet-config.c
 * @brief tool to access and manipulate GNUnet configuration files
 * @author Christian Grothoff
 */
#include "platform.h"
#include "gnunet_util_lib.h"


/**
 * Name of the section
 */
static char *section;

/**
 * Name of the option
 */
static char *option;

/**
 * Value to set
 */
static char *value;

/**
 * Backend to check if the respective plugin is
 * loadable. NULL if no check is to be performed.
 * The value is the "basename" of the plugin to load.
 */
static char *backend_check;

/**
 * Treat option as a filename.
 */
static int is_filename;

/**
 * Whether to show the sections.
 */
static int list_sections;

/**
 * Return value from 'main'.
 */
static int global_ret;

/**
 * Should we generate a configuration file that is clean and
 * only contains the deltas to the defaults?
 */
static int rewrite;


/**
 * Print each option in a given section.
 *
 * @param cls closure
 * @param section name of the section
 * @param option name of the option
 * @param value value of the option
 */
static void
print_option (void *cls,
              const char *section,
              const char *option,
              const char *value)
{
  const struct GNUNET_CONFIGURATION_Handle *cfg = cls;

  (void) section;
  if (is_filename)
  {
    char *value_fn;
    char *fn;

    GNUNET_assert (GNUNET_OK ==
                   GNUNET_CONFIGURATION_get_value_filename (cfg,
                                                            section,
                                                            option,
                                                            &value_fn));
    fn = GNUNET_STRINGS_filename_expand (value_fn);
    if (NULL == fn)
      fn = value_fn;
    else
      GNUNET_free (value_fn);
    fprintf (stdout, "%s = %s\n", option, fn);
    GNUNET_free (fn);
  }
  else
  {
    fprintf (stdout, "%s = %s\n", option, value);
  }
}


/**
 * Print out given section name.
 *
 * @param cls unused
 * @param section a section in the configuration file
 */
static void
print_section_name (void *cls, const char *section)
{
  (void) cls;
  fprintf (stdout, "%s\n", section);
}


/**
 * Main function that will be run by the scheduler.
 *
 * @param cls closure
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving,
 *                                                     can be NULL!)
 * @param cfg configuration
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *cfg)
{
  struct GNUNET_CONFIGURATION_Handle *out = NULL;
  struct GNUNET_CONFIGURATION_Handle *diff = NULL;
  char *cfg_fn;

  (void) cls;
  (void) args;
  if (NULL != backend_check)
  {
    char *name;

    GNUNET_asprintf (&name, "libgnunet_plugin_%s", backend_check);
    global_ret = (GNUNET_OK == GNUNET_PLUGIN_test (name)) ? 0 : 77;
    GNUNET_free (name);
    return;
  }
  if (rewrite)
  {
    struct GNUNET_CONFIGURATION_Handle *def;

    def = GNUNET_CONFIGURATION_create ();
    if (GNUNET_OK != GNUNET_CONFIGURATION_load (def, NULL))
    {
      fprintf (stderr, _ ("failed to load configuration defaults"));
      global_ret = 1;
      return;
    }
    diff = GNUNET_CONFIGURATION_get_diff (def, cfg);
    cfg = diff;
  }
  if (((! rewrite) && (NULL == section)) || list_sections)
  {
    if (! list_sections)
    {
      fprintf (stderr,
               _ ("%s or %s argument is required\n"),
               "--section",
               "--list-sections");
      global_ret = 1;
    }
    else
    {
      fprintf (stderr, _ ("The following sections are available:\n"));
      GNUNET_CONFIGURATION_iterate_sections (cfg, &print_section_name, NULL);
    }
    goto cleanup;
  }

  if ((NULL != section) && (NULL == value))
  {
    if (NULL == option)
    {
      GNUNET_CONFIGURATION_iterate_section_values (cfg,
                                                   section,
                                                   &print_option,
                                                   (void *) cfg);
    }
    else
    {
      if (is_filename)
      {
        if (GNUNET_OK != GNUNET_CONFIGURATION_get_value_filename (cfg,
                                                                  section,
                                                                  option,
                                                                  &value))
        {
          GNUNET_log_config_missing (GNUNET_ERROR_TYPE_ERROR, section, option);
          global_ret = 3;
          goto cleanup;
        }
      }
      else
      {
        if (GNUNET_OK != GNUNET_CONFIGURATION_get_value_string (cfg,
                                                                section,
                                                                option,
                                                                &value))
        {
          GNUNET_log_config_missing (GNUNET_ERROR_TYPE_ERROR, section, option);
          global_ret = 3;
          goto cleanup;
        }
      }
      fprintf (stdout, "%s\n", value);
    }
  }
  else if (NULL != section)
  {
    if (NULL == option)
    {
      fprintf (stderr, _ ("--option argument required to set value\n"));
      global_ret = 1;
      goto cleanup;
    }
    out = GNUNET_CONFIGURATION_dup (cfg);
    GNUNET_CONFIGURATION_set_value_string (out, section, option, value);
  }
  cfg_fn = NULL;
  if (NULL == cfgfile)
  {
    const char *xdg = getenv ("XDG_CONFIG_HOME");
    if (NULL != xdg)
      GNUNET_asprintf (&cfg_fn,
                       "%s%s%s",
                       xdg,
                       DIR_SEPARATOR_STR,
                       GNUNET_OS_project_data_get ()->config_file);
    else
      cfg_fn = GNUNET_strdup (GNUNET_OS_project_data_get ()->user_config_file);
    cfgfile = cfg_fn;
  }
  if ((NULL != diff) || (NULL != out))
  {
    if (GNUNET_OK !=
        GNUNET_CONFIGURATION_write ((NULL == out) ? diff : out, cfgfile))
      global_ret = 2;
  }
  GNUNET_free_non_null (cfg_fn);
  if (NULL != out)
    GNUNET_CONFIGURATION_destroy (out);
cleanup:
  if (NULL != diff)
    GNUNET_CONFIGURATION_destroy (diff);
}


/**
 * Program to manipulate configuration files.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 ok, 1 on error
 */
int
main (int argc, char *const *argv)
{
  struct GNUNET_GETOPT_CommandLineOption options[] =
  { GNUNET_GETOPT_option_flag (
      'f',
      "filename",
      gettext_noop ("interpret option value as a filename (with $-expansion)"),
      &is_filename),
    GNUNET_GETOPT_option_exclusive (GNUNET_GETOPT_option_string (
                                      'b',
                                      "supported-backend",
                                      "BACKEND",
                                      gettext_noop (
                                        "test if the current installation supports the specified BACKEND"),
                                      &backend_check)),
    GNUNET_GETOPT_option_string ('s',
                                 "section",
                                 "SECTION",
                                 gettext_noop (
                                   "name of the section to access"),
                                 &section),
    GNUNET_GETOPT_option_string ('o',
                                 "option",
                                 "OPTION",
                                 gettext_noop ("name of the option to access"),
                                 &option),
    GNUNET_GETOPT_option_string ('V',
                                 "value",
                                 "VALUE",
                                 gettext_noop ("value to set"),
                                 &value),
    GNUNET_GETOPT_option_flag ('S',
                               "list-sections",
                               gettext_noop (
                                 "print available configuration sections"),
                               &list_sections),
    GNUNET_GETOPT_option_flag (
      'w',
      "rewrite",
      gettext_noop (
        "write configuration file that only contains delta to defaults"),
      &rewrite),
    GNUNET_GETOPT_OPTION_END };
  int ret;

  if (GNUNET_OK != GNUNET_STRINGS_get_utf8_args (argc, argv, &argc, &argv))
    return 2;

  ret =
    GNUNET_PROGRAM_run (argc,
                        argv,
                        "gnunet-config [OPTIONS]",
                        gettext_noop ("Manipulate GNUnet configuration files"),
                        options,
                        &run,
                        NULL);
  GNUNET_free ((void *) argv);
  if (GNUNET_OK == ret)
    return global_ret;
  return ret;
}

/* end of gnunet-config.c */
